src/utils/connect.js

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import * as React from 'react';
+import { connect } from 'react-redux';
+import type {} from 'react-redux';
+import type { Dispatch, State } from '../types/store';
+
+type MapStateToProps<OwnProps: Object, StateProps: Object> = (
+  state: State,
+  ownProps: OwnProps
+) => StateProps;
+
+type MapDispatchToProps<OwnProps: Object, DispatchProps: Object> =
+  | ((dispatch: Dispatch, ownProps: OwnProps) => DispatchProps)
+  | DispatchProps;
+
+type MergeProps<
+  StateProps,
+  DispatchProps: Object,
+  OwnProps: Object,
+  Props: Object
+> = (
+  stateProps: StateProps,
+  dispatchProps: DispatchProps,
+  ownProps: OwnProps
+) => Props;
+
+type ConnectOptions = {
+  pure?: boolean,
+  areStatesEqual?: boolean,
+  areOwnPropsEqual?: boolean,
+  areStatePropsEqual?: boolean,
+  areMergedPropsEqual?: boolean,
+  storeKey?: boolean,
+  withRef?: boolean,
+};
+
+export type ExplicitConnectOptions<
+  OwnProps: Object,
+  StateProps: Object,
+  DispatchProps: Object
+> = {
+  mapStateToProps?: MapStateToProps<OwnProps, StateProps>,
+  mapDispatchToProps?: MapDispatchToProps<OwnProps, DispatchProps>,
+  mergeProps?: MergeProps<
+    StateProps,
+    DispatchProps,
+    OwnProps,
+    {
+      ...OwnProps,
+      ...StateProps,
+      ...DispatchProps,
+    }
+  >,
+  options?: ConnectOptions,
+  component:
+    | React.ComponentType<{|
+        ...OwnProps,
+        ...StateProps,
+        ...DispatchProps,
+      |}>
+    | React.ComponentType<{
+        ...OwnProps,
+        ...StateProps,
+        ...DispatchProps,
+      }>,
```

Comment:
could you add a comment explaining why we need these 2 flavors ?
(as said earlier, you'll likely have to add `$ReadOnly` here too)

Reply:
I don't anymore.